runtime/src/accounts_index.rs

```diff
@@ -1,38 +1,276 @@
 use solana_sdk::{clock::Slot, pubkey::Pubkey};
+use std::ops::{
+    Bound,
+    Bound::{Excluded, Included, Unbounded},
+};
 use std::sync::atomic::{AtomicU64, Ordering};
 use std::{
     collections::{BTreeMap, HashMap, HashSet},
     ops::RangeBounds,
-    sync::{RwLock, RwLockReadGuard},
+    sync::{Arc, RwLock},
 };
 
+const ITER_BATCH_SIZE: usize = 1000;
+
+rental! {
+    mod owning_lock {
+        use std::sync::{Arc, RwLock, RwLockReadGuard, RwLockWriteGuard};
+
+        #[rental(deref_suffix)]
+        pub struct OwningReadGuard<T>
+        where
+            T: 'static,
+        {
+            lock: Arc<RwLock<T>>,
+            guard: RwLockReadGuard<'lock, T>,
+        }
+
+        #[rental(deref_mut_suffix)]
+        pub struct OwningWriteGuard<T>
+        where
+            T: 'static,
+        {
+            lock: Arc<RwLock<T>>,
+            guard: RwLockWriteGuard<'lock, T>,
+        }
+    }
+}
+
+use owning_lock::{OwningReadGuard, OwningWriteGuard};
+
 pub type SlotList<T> = Vec<(Slot, T)>;
 pub type SlotSlice<'s, T> = &'s [(Slot, T)];
 pub type Ancestors = HashMap<Slot, usize>;
 
 pub type RefCount = u64;
-type AccountMapEntry<T> = (AtomicU64, RwLock<SlotList<T>>);
 pub type AccountMap<K, V> = BTreeMap<K, V>;
 
+#[derive(Clone, Debug)]
+pub struct AccountMapEntry<T> {
+    ref_count: Arc<AtomicU64>,
+    pub slot_list: Arc<RwLock<SlotList<T>>>,
+}
```

Comment:
I wonder we could reduce Arc to 1 from 2 by line this, it's a bit hideous but maybe worth effort? (we should be keen the size of AcccontIndex memory footprint (Arc costs 2-word memory for each, so we're increasing entry size by `4 * usize` here per `Account`.) and reduced general update cost from updating Arc twice everytime accessed).

```rust
pub struct AccountMapEntryInner<T> {
    ref_count: AtomicU64,
    pub slot_list: RwLock<SlotList<T>>,
}
type AccountMapEntry<T> = Arc<AccountMapEntryInner>;
```

Reply:
@ryoqun  this is doable, but I think it makes the self-referential struct a little uglier because now you have to wrap the entire `AccountsMapEntryInner` instead of just the lock inside the `self-reference` tag. Here's an example:  https://github.com/solana-labs/solana/pull/12787/commits/50a9b2647d6a47c0758b78e277731e5731accd5c

What do you think?

Reply:
@ryoqun this is much cleaner with the ouroboros crate here: https://github.com/solana-labs/solana/pull/12787/commits/970b2c78aa1a04a980500591fd2cc5cbc6af6be5